pallets/amm/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+        }
+
+        // Related to the underlying formula used to evaluate swaps. Calculates `y` (quote) amount
+        // and the different between the previous `y` amount based on the provided amount of
+        // `x` (base).
+        //
+        // let y = (k * (4*A*k + k - 4*A*x)) / (4 * (A*k + x))
+        fn calculate_y(x: Balance, pool: &Pool) -> Result<YEvaluation, DispatchError> {
+            let k = Self::total_assets(pool)?;
+            let f = || {
+                let ak = Self::amplification_coeficient_mul(k)?;
+                let _4ax = 4u128.checked_mul(Self::amplification_coeficient_mul(x)?)?;
+                let _4ak = 4u128.checked_mul(Self::amplification_coeficient_mul(k)?)?;
+                let dividend = k.checked_mul(_4ak.checked_add(k)?.checked_sub(_4ax)?)?;
+                let divisor = 4u128.checked_mul(ak.checked_add(x)?)?;
+                dividend.checked_div(divisor)
+            };
+            let y = f().ok_or_else::<DispatchError, _>(|| ArithmeticError::Overflow.into())?;
+            let [greater, lesser] = if pool.amount_quote > y {
+                [pool.amount_quote, y]
+            } else {
+                [y, pool.amount_quote]
+            };
+            let y_diff = greater
+                .checked_sub(lesser)
+                .ok_or_else::<DispatchError, _>(|| ArithmeticError::Underflow.into())?;
+            Ok(YEvaluation { y, y_diff })
+        }
+
+        // Increases the internal pool counter by 1 and returns the id that should
+        // be attached to a new stored pool.
+        fn increase_pools_counter() -> Result<T::PoolId, DispatchError> {
+            let curr = <PoolsCounter<T>>::get();
+            <PoolsCounter<T>>::try_mutate(|n| {
+                let opt = n.checked_add(&T::PoolId::from(1u8));
+                *n = opt.ok_or_else::<DispatchError, _>(|| ArithmeticError::Overflow.into())?;
+                Ok::<_, DispatchError>(())
+            })?;
+            Ok(curr)
+        }
+
+        fn mutate_pool<F>(pool_id: &T::PoolId, cb: F) -> Result<(), DispatchError>
+        where
+            F: FnOnce(&mut Pool),
+        {
+            Pools::<T>::try_mutate(pool_id, |opt| {
+                if let Some(market) = opt {
+                    cb(market);
+                    return Ok(());
+                }
+                Err(Error::<T>::PoolDoesNotExist.into())
+            })
+        }
+
+        fn pool(pool_id: &T::PoolId) -> Result<Pool, DispatchError> {
+            Pools::<T>::get(pool_id).ok_or_else(|| Error::<T>::PoolDoesNotExist.into())
+        }
+
+        // Sum of the two pool assets
+        fn total_assets(pool: &Pool) -> Result<Balance, DispatchError> {
+            pool.amount_base
+                .checked_add(pool.amount_quote)
+                .ok_or_else(|| ArithmeticError::Overflow.into())
+        }
+    }
+
+    #[pallet::storage]
+    pub type ExchangeRate<T: Config> = StorageValue<_, Rate, ValueQuery>;
```

Comment:
Just to show how the exchange rate would be used to shape the curve.

Both `amm` and `loans` use stored `exchange_rate` values so it is probably better to created a common pallet instead of duplicating error-probe code.